### PR TITLE
Strict map

### PR DESCRIPTION
**Pull request by kazu-yamamoto**
Using strict map and int map just in case.

### CODE REVIEW

File: dnsext-iterative/DNS/Iterative/Query/LocalZone.hs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 module DNS.Iterative.Query.LocalZone where
 
 -- GHC packages
-import Data.Map (Map)
+import Data.Map.Strict (Map)
```
Comment (khibino) [inline dnsext-iterative/DNS/Iterative/Query/LocalZone.hs:4]: `Data.Map.Map` and `Data.Map.Strict.Map` are the same type, is this change necessary?

File: dnsext-iterative/DNS/Iterative/Query/StubZone.hs [hunk 2]
```diff
@@ -1,7 +1,7 @@
 module DNS.Iterative.Query.StubZone where
 
 -- GHC packages
-import Data.Map (Map)
+import Data.Map.Strict (Map)
```
Comment (khibino) [inline dnsext-iterative/DNS/Iterative/Query/StubZone.hs:4]: > `Data.Map.Map` and `Data.Map.Strict.Map` are the same type, is this change necessary?

Same point as the LocalZone-module one here.

**Review by khibino — COMMENTED** [review]
`Map` and `IntMap` change the operation to strict,
since the type is the same, why not leave the import statement of the type unchanged?

File: dnsext-iterative/DNS/Iterative/Query/LocalZone.hs (hunk 1 above)
Comment (kazu-yamamoto) [reply dnsext-iterative/DNS/Iterative/Query/LocalZone.hs:4]: When I do `grep -r Data.Map */DNS | grep -v Data.Map.Strict`, this change ensures that safe lines are not displayed.

**Comment by kazu-yamamoto** [comment]
> `Map` and `IntMap` change the operation to strict,
> since the type is the same, why not leave the import statement of the type unchanged?

When we read the following code, we need to have knowledge where operations of `Data.Map.Strict` can be applied to `Data.Map.Map`.

```haskell
import Data.Map (Map)
import qualified Data.Map.Strict as M
```

However, we can read the following code without such knowledge and surprise.

```haskell
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as M
```

**Comment by khibino** [comment]
> When we read the following code, we need to have knowledge where operations of Data.Map.Strict can be applied to Data.Map.Map.

But knowing that `Data.Map.Map` and `Data.Map.Strict.Map` are the same type,
the import statement of `import Data.Map.Strict (Map)` will seem strange to you.

I think it would be better to require that the person who works with `Data.Map.Strict` knows that the type does not change in strict operations.

**Comment by kazu-yamamoto** [comment]
Suppose we adopt the following style:

```haskell
import Data.Map (Map)
import qualified Data.Map.Strict as M
```

How we can find wrong operations from `Data.Map` without *noise*?

**Comment by khibino** [comment]
I believe that whether using lazy operations on Data.Map is wrong or not cannot be determined without considering the intended usage at each site.
I don't think it's generally appropriate to make this determination based solely on `grep` results.

**Comment by kazu-yamamoto** [comment]
I don't know examples where lazy Map is better than strict one.
Rather, lazy Map is a main source of memory leak.
Honestly, I'm not a fan of of lazy evaluation.

`Data.Map` should be strict on entries by default.
In this sense, `Data.Map.Strict` has a bad name because of historical reasons.

Even if you feel `Data.Map.Strict (Map)` is strange, I can say I don't feel it is strange.
It's matter of apple or orange.

I would like to have a simple method to detect usage of lazy operations.

**Comment by khibino** [comment]
Considering cases where a lazy Map is effective,
it would be when the values are thunks with large computational complexity,
but which do not increase in size.
A simple example would be Dynamic Programming with non-integer indices.

If there are cases where strict is not used,
it concerns me that there are two ways to import the Map type:
`import Data.Map (Map)` and `import Data.Map.Strict (Map)`.
I would like to do imports of the same definition in the same import statement.

**Comment by kazu-yamamoto** [comment]
After private discussion, this PR has been merged.
